File: implementations/csharp/Provekit.ProofEnvelope/Proof.cs
// SPDX-License-Identifier: Apache-2.0
//
// .proof envelope builder. Per RFC 8949 §4.2.1 + the .proof spec
// (protocol/specs/2026-04-30-proof-file-format.md):
//
//   1. Build the unsigned body as a CBOR map with keys sorted by
//      bytewise lex order of their CBOR-encoded form.
//   2. Ed25519-sign the unsigned-body bytes.
//   3. Re-emit the body with the signature added; keys re-sort
//      automatically.
//   4. BLAKE3-512 the final bytes; the full self-identifying string
//      `"blake3-512:<128 hex>"` IS the catalog CID.
//
// The `members` map key is the embedded envelope's own CID; the value
// is its canonical bytes (JCS-JSON for memento envelopes per the
// memento envelope grammar) wrapped as a CBOR byte string.

using Provekit.Canonicalizer;

namespace Provekit.ProofEnvelope;

public sealed class ProofEnvelopeInput
{
    public string Name { get; init; } = "";
    public string Version { get; init; } = "";

    /// <summary>
    /// Map from member CID (full self-identifying string,
    /// e.g. "blake3-512:abc...") to that member's canonical bytes
    /// (JCS-JSON bytes for memento envelopes).
    /// </summary>
    public IReadOnlyDictionary<string, byte[]> Members { get; init; } =
        new Dictionary<string, byte[]>();

    /// <summary>CID of the signer's public-key memento (or any resolvable CID).</summary>
    public string SignerCid { get; init; } = "";

    /// <summary>Ed25519 seed bytes (32). Deterministic signing.</summary>
    public byte[] SignerSeed { get; init; } = Array.Empty<byte>();

    /// <summary>ISO-8601 string with millisecond precision, trailing 'Z'.</summary>
    public string DeclaredAt { get; init; } = "";
}

public sealed class ProofEnvelopeOutput
{
    /// <summary>CBOR bytes of the signed catalog. Hash of these bytes IS the CID.</summary>
    public required byte[] Bytes { get; init; }

    /// <summary>Full self-identifying CID, e.g. "blake3-512:&lt;128 hex&gt;".</summary>
    public required string Cid { get; init; }
}

internal sealed class CborPair
{
    public required byte[] KeyCbor { get; init; }
    public required byte[] ValueCbor { get; init; }
}

public static class Proof
{
    public static ProofEnvelopeOutput Build(ProofEnvelopeInput input)
    {
        // Step 1: encode unsigned body with sorted keys.
        var unsignedPairs = BodyPairsUnsigned(input);
        var unsignedBytes = new List<byte>();
        EmitSortedMap(unsignedBytes, unsignedPairs);
        var unsignedArr = unsignedBytes.ToArray();

        // Step 2: Ed25519-sign the unsigned bytes.
        var sig = Sign.Ed25519SignWithSeed(input.SignerSeed, unsignedArr);

        // Step 3: re-emit with signature added; keys re-sort automatically.
        var signedPairs = BodyPairsUnsigned(input);
        signedPairs.Add(MakeBytesPair("signature", sig));
        var finalBytes = new List<byte>();
        EmitSortedMap(finalBytes, signedPairs);
        var finalArr = finalBytes.ToArray();

        // Step 4: filename CID = full self-identifying BLAKE3-512.
        var cid = Hash.Blake3_512(finalArr);
        return new ProofEnvelopeOutput { Bytes = finalArr, Cid = cid };
    }

    private static List<CborPair> BodyPairsUnsigned(ProofEnvelopeInput input)
    {
        return new List<CborPair>
        {
            MakeStringPair("kind", "catalog"),
            MakeStringPair("name", input.Name),
            MakeStringPair("version", input.Version),
            MakeMembersPair("members", input.Members),
            MakeStringPair("signer", input.SignerCid),
            MakeStringPair("declaredAt", input.DeclaredAt),
        };
    }

    private static byte[] EncodeKey(string key)
    {
        var k = new List<byte>();
        Cbor.EncodeTstr(k, key);
        return k.ToArray();
    }

    private static CborPair MakeStringPair(string key, string value)
    {
        var v = new List<byte>();
        Cbor.EncodeTstr(v, value);
        return new CborPair { KeyCbor = EncodeKey(key), ValueCbor = v.ToArray() };
    }

    private static CborPair MakeBytesPair(string key, byte[] value)
    {
        var v = new List<byte>();
        Cbor.EncodeBstr(v, value);
        return new CborPair { KeyCbor = EncodeKey(key), ValueCbor = v.ToArray() };
    }

    private static CborPair MakeMembersPair(string key, IReadOnlyDictionary<string, byte[]> members)
    {
        var pairs = new List<CborPair>();
        foreach (var kv in members)
        {
            pairs.Add(MakeBytesPair(kv.Key, kv.Value));
        }
        var valueBytes = new List<byte>();
        EmitSortedMap(valueBytes, pairs);
        return new CborPair { KeyCbor = EncodeKey(key), ValueCbor = valueBytes.ToArray() };
    }

    private static void EmitSortedMap(List<byte> output, List<CborPair> pairs)
    {
        // Sort by bytewise lex order of the CBOR-encoded key form (RFC
        // 8949 §4.2.1).
        pairs.Sort((a, b) => CompareBytewise(a.KeyCbor, b.KeyCbor));
        Cbor.EncodeMapHead(output, (ulong)pairs.Count);
        foreach (var p in pairs)
        {
            output.AddRange(p.KeyCbor);
            output.AddRange(p.ValueCbor);
        }
    }

    private static int CompareBytewise(byte[] a, byte[] b)
    {
        var n = Math.Min(a.Length, b.Length);
        for (var i = 0; i < n; i++)
        {
            var d = a[i].CompareTo(b[i]);
            if (d != 0) return d;
        }
        return a.Length.CompareTo(b.Length);
    }
}

File: implementations/rust/provekit-proof-envelope/src/proof.rs
// SPDX-License-Identifier: Apache-2.0
//
// .proof envelope builder. Per RFC 8949 §4.2.1 + the .proof spec
// (protocol/specs/2026-04-30-proof-file-format.md):
//
//   1. Build the unsigned body as a CBOR map with keys sorted by
//      bytewise lex order of their CBOR-encoded form.
//   2. ed25519-sign the unsigned-body bytes.
//   3. Re-emit the body with the signature added; keys re-sort
//      automatically (the new "signature" key slots in by lex order).
//   4. BLAKE3-512 the final bytes; the full self-identifying string
//      `"blake3-512:<128 hex>"` IS the catalog CID.
//
// The `members` map key is the embedded envelope's own CID, and the
// value is its canonical bytes (JCS-JSON for memento envelopes per
// the memento envelope grammar) wrapped as a CBOR byte string.

use std::collections::BTreeMap;

use provekit_canonicalizer::blake3_512_of;

use crate::cbor::{
    cbor_encode_bstr, cbor_encode_map_head, cbor_encode_tstr, cbor_encode_uint, CborMajor,
};
use crate::sign::{ed25519_sign_with_seed, Ed25519Seed};

#[derive(Debug, Clone)]
pub struct ProofEnvelopeInput {
    pub name: String,
    pub version: String,
    /// Optional CID of the compiled binary this proof verifies.
    /// When present, the framework checks that the running binary's
    /// hash matches before trusting any claims in this bundle.
    pub binary_cid: Option<String>,
    /// Optional metadata: key-value map for tooling and diagnostics.
    /// Included in the signed payload (tamper-evident) but explicitly
    /// NON-NORMATIVE: verifiers MUST NOT use metadata for logic.
    pub metadata: Option<BTreeMap<String, String>>,
    /// Map from member CID (full self-identifying string form,
    /// e.g. `"blake3-512:abc..."`) to that member's canonical bytes
    /// (JCS-JSON bytes for memento envelopes).
    pub members: BTreeMap<String, Vec<u8>>,
    /// Identity carried into the envelope's `signer` field. Two valid
    /// shapes per the memento-envelope grammar:
    ///   1. `blake3-512:<hex>` - CID resolving to a pubkey memento
    ///      (canonical, used by `.proof` catalogs)
    ///   2. `ed25519:<base64-pubkey>` - inline self-identifying form
    ///      (used by memento envelopes that don't pin a separate
    ///      pubkey memento, e.g. fixtures and standalone declarations)
    ///
    /// The struct field name is preserved for backwards compatibility;
    /// either form is accepted by the wire format.
    pub signer_cid: String,
    /// Ed25519 seed bytes; deterministic signing for tests/demos.
    pub signer_seed: Ed25519Seed,
    /// ISO-8601 string with millisecond precision and trailing 'Z'.
    pub declared_at: String,
}

#[derive(Debug, Clone)]
pub struct ProofEnvelopeOutput {
    /// CBOR bytes of the signed catalog. Hash of these bytes IS the CID.
    pub bytes: Vec<u8>,
    /// Full self-identifying CID, e.g. `"blake3-512:<128 hex>"`.
    pub cid: String,
}

// One CBOR (key, value) pair, with its key pre-encoded so the
// outer map can sort by bytewise CBOR-encoded-key form.
struct CborPair {
    key_cbor: Vec<u8>,
    value_cbor: Vec<u8>,
}

fn encode_key(key: &str) -> Vec<u8> {
    let mut k = Vec::with_capacity(1 + key.len());
    cbor_encode_tstr(&mut k, key);
    k
}

fn make_string_pair(key: &str, value: &str) -> CborPair {
    let mut v = Vec::with_capacity(1 + value.len());
    cbor_encode_tstr(&mut v, value);
    CborPair {
        key_cbor: encode_key(key),
        value_cbor: v,
    }
}

fn make_bytes_pair(key: &str, value: &[u8]) -> CborPair {
    let mut v = Vec::with_capacity(1 + value.len());
    cbor_encode_bstr(&mut v, value);
    CborPair {
        key_cbor: encode_key(key),
        value_cbor: v,
    }
}

fn make_members_pair(key: &str, members: &BTreeMap<String, Vec<u8>>) -> CborPair {
    // Encode as { tstr(cid) => bstr(envelope-bytes) }, sort by
    // bytewise CBOR-encoded-key form.
    let mut pairs: Vec<CborPair> = members
        .iter()
        .map(|(cid, bytes)| make_bytes_pair(cid, bytes))
        .collect();
    let mut value_cbor = Vec::new();
    emit_sorted_map(&mut value_cbor, &mut pairs);
    CborPair {
        key_cbor: encode_key(key),
        value_cbor,
    }
}

fn emit_sorted_map(out: &mut Vec<u8>, pairs: &mut [CborPair]) {
    pairs.sort_by(|a, b| a.key_cbor.cmp(&b.key_cbor));
    cbor_encode_map_head(out, pairs.len() as u64);
    for p in pairs.iter() {
        out.extend_from_slice(&p.key_cbor);
        out.extend_from_slice(&p.value_cbor);
    }
}

fn body_pairs_unsigned(input: &ProofEnvelopeInput) -> Vec<CborPair> {
    let mut pairs = vec![
        make_string_pair("kind", "catalog"),
        make_string_pair("name", &input.name),
        make_string_pair("version", &input.version),
        make_members_pair("members", &input.members),
        make_string_pair("signer", &input.signer_cid),
        make_string_pair("declaredAt", &input.declared_at),
    ];
    if let Some(ref bcid) = input.binary_cid {
        pairs.push(make_string_pair("binaryCid", bcid));
    }
    if let Some(ref meta) = input.metadata {
        let mut meta_pairs: Vec<CborPair> =
            meta.iter().map(|(k, v)| make_string_pair(k, v)).collect();
        let mut meta_cbor = Vec::new();
        emit_sorted_map(&mut meta_cbor, &mut meta_pairs);
        pairs.push(CborPair {
            key_cbor: encode_key("metadata"),
            value_cbor: meta_cbor,
        });
    }
    pairs
}

pub fn build_proof_envelope(input: &ProofEnvelopeInput) -> ProofEnvelopeOutput {
    // Step 1: encode unsigned body with sorted keys.
    let mut unsigned_pairs = body_pairs_unsigned(input);
    let mut unsigned_bytes = Vec::new();
    emit_sorted_map(&mut unsigned_bytes, &mut unsigned_pairs);

    // Step 2: ed25519-sign the unsigned bytes.
    let sig = ed25519_sign_with_seed(&input.signer_seed, &unsigned_bytes);

    // Step 3: re-emit with signature added; keys re-sort automatically.
    let mut signed_pairs = body_pairs_unsigned(input);
    signed_pairs.push(make_bytes_pair("signature", &sig));
    let mut final_bytes = Vec::new();
    emit_sorted_map(&mut final_bytes, &mut signed_pairs);

    // Silence dead-code from imports we keep for callers.
    let _ = (cbor_encode_uint, CborMajor::UnsignedInt);

    // Step 4: filename CID = full self-identifying BLAKE3-512.
    let cid = blake3_512_of(&final_bytes);
    ProofEnvelopeOutput {
        bytes: final_bytes,
        cid,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_minimal_proof_round_trips() {
        let mut members = BTreeMap::new();
        members.insert(
            "blake3-512:aa".to_string(),
            b"{\"hello\":\"world\"}".to_vec(),
        );
        let input = ProofEnvelopeInput {
            name: "@x/y".to_string(),
            version: "0.0.1".to_string(),
            binary_cid: None,
            metadata: None,
            members,
            signer_cid: "blake3-512:bb".to_string(),
            signer_seed: [0x11; 32],
            declared_at: "2026-04-30T00:00:00.000Z".to_string(),
        };
        let out = build_proof_envelope(&input);
        assert!(out.cid.starts_with("blake3-512:"));
        // First byte is map head with 7 entries (kind, name, version, members,
        // signer, declaredAt, signature) = 0xa7.
        assert_eq!(out.bytes[0], 0xa7);
    }
}
